src/de/raw.rs

```diff
@@ -1244,118 +1034,80 @@ enum BinaryDeserializationStage {
-            CodeWithScopeDeserializationStage::Code => seed
-                .deserialize(FieldDeserializer {
-                    field_name: "$code",
-                })
-                .map(Some),
-            CodeWithScopeDeserializationStage::Scope => seed
-                .deserialize(FieldDeserializer {
-                    field_name: "$scope",
-                })
-                .map(Some),
-            CodeWithScopeDeserializationStage::Done => Ok(None),
-        }
+        let field_name = match self.stage {
+            CodeWithScopeDeserializationStage::Code => "$code",
+            CodeWithScopeDeserializationStage::Scope => "$scope",
+            CodeWithScopeDeserializationStage::Done => return Ok(None),
+        };
+        seed.deserialize(FieldDeserializer { field_name }).map(Some)
     }
 
-    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value>
+    fn next_value_seed<V>(&mut self, seed: V) -> std::result::Result<V::Value, Self::Error>
     where
         V: serde::de::DeserializeSeed<'de>,
     {
-        seed.deserialize(&mut *self.deserializer)
-    }
-}
-
-struct CodeWithScopeDeserializer<'de, 'a> {
-    root_deserializer: &'a mut Deserializer<'de>,
-    stage: CodeWithScopeDeserializationStage,
-    hint: DeserializerHint,
-    length_remaining: i32,
-}
-
-impl<'de, 'a> CodeWithScopeDeserializer<'de, 'a> {
-    fn new(root_deserializer: &'a mut Deserializer<'de>, hint: DeserializerHint, len: i32) -> Self {
-        Self {
-            root_deserializer,
-            stage: CodeWithScopeDeserializationStage::Code,
-            hint,
-            length_remaining: len,
-        }
-    }
-
-    /// Executes a closure that reads from the BSON bytes and returns an error if the number of
-    /// bytes read exceeds length_remaining.
-    ///
-    /// A mutable reference to this [`CodeWithScopeDeserializer`] is passed into the closure.
-    fn read<F, O>(&mut self, f: F) -> Result<O>
-    where
-        F: FnOnce(&mut Self) -> Result<O>,
-    {
-        let start_bytes = self.root_deserializer.bytes.bytes_read();
-        let out = f(self)?;
-        let bytes_read = self.root_deserializer.bytes.bytes_read() - start_bytes;
-        self.length_remaining -= bytes_read as i32;
-
-        if self.length_remaining < 0 {
-            return Err(Error::custom("length of CodeWithScope too short"));
-        }
-        Ok(out)
+        let value = seed.deserialize(&*self)?;
+        self.stage = match self.stage {
+            CodeWithScopeDeserializationStage::Code => CodeWithScopeDeserializationStage::Scope,
+            CodeWithScopeDeserializationStage::Scope => CodeWithScopeDeserializationStage::Done,
+            CodeWithScopeDeserializationStage::Done => return Err(Error::EndOfStream),
+        };
+        Ok(value)
     }
 }
 
-impl<'de, 'a, 'b> serde::de::Deserializer<'de> for &'b mut CodeWithScopeDeserializer<'de, 'a> {
+impl<'a, 'de> serde::de::Deserializer<'de> for &'a CodeWithScopeAccess<'de> {
```

Comment:
For the various `*Access` types that needed to be updated, I found it simpler to have them implement both the `Access` and `Deserializer` traits rather than splitting across two types.